<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="epi2me2r">
<title>epi2me2r vignette • epi2me2r</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="epi2me2r vignette">
<meta property="og:description" content="epi2me2r">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">epi2me2r</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/epi2me2r-vignette.html">epi2me2r vignette</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="epi2me2r-vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>epi2me2r vignette</h1>
                        <h4 data-toc-skip class="author">Maggie Weinroth</h4>
            
      
      
      <div class="d-none name"><code>epi2me2r-vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mweinroth.github.io/epi2me2r/">epi2me2r</a></span><span class="op">)</span></code></pre></div>
<p>epi2me2r is meant to full a fully automated methods to take raw csv of AMRA and WIMP output from the EPI2ME pipeline from <a href="https://nanoporetech.com/" class="external-link">Oxford Nanopore</a> to quickly convert it into common R package formats, namely <a href="https://joey711.github.io/phyloseq/" class="external-link">phyloseq</a> and <a href="https://www.bioconductor.org/packages/release/bioc/html/metagenomeSeq.html" class="external-link">metagenomeSeq</a>.</p>
<p>There are three main types of functions in epi2me2r:</p>
<ul>
<li>
<strong>fully automated</strong>: these functions take minimal input (usually just raw csv files and metadata) and produce either phyloseq and metagenomicSeq objects for downstream analysis:
<ul>
<li><code>raw_amr_to_phyloseq</code></li>
<li><code>raw_amr_to_metagenomseq</code></li>
<li><code>raw_wimp_to_phyloseq</code></li>
<li><code>raw_wimp_to_metagenomeseq</code></li>
</ul>
</li>
<li>
<strong>step-by-step</strong>: If you are looking for just a portion of the data you can these functions to generation only what you need:
<ul>
<li><code>read_in_amr_files</code></li>
<li><code>read_in_wimp_files</code></li>
<li><code>generate_amr_taxonomy</code></li>
<li><code>generate_wimp_taxonomy</code></li>
</ul>
</li>
<li>
<strong>other</strong>: An additional function we created but does not fit into the main workflows is <code>amr_read_taxonomy</code>. This function reads in both AMR and WIMP raw data and adds the taxonomic infromation to the AMR gene if available.</li>
</ul>
<div class="section level2">
<h2 id="preperation-of-metadata-file">preperation of metadata file<a class="anchor" aria-label="anchor" href="#preperation-of-metadata-file"></a>
</h2>
<p>Prior to starting, making sure the metadata file is formatted appropriately and in the correct format will ensure your data is imported correctly. You can use one metadata file for both your AMR and WIMP samples or separate. Both options are shown below.</p>
<div class="section level4">
<h4 id="combo-metadata-file">Combo metadata file<a class="anchor" aria-label="anchor" href="#combo-metadata-file"></a>
</h4>
<p>This file has <strong>4</strong> required columns that must been entered as seen below:</p>
<ul>
<li>
<code>amr_filename</code> : the original amr file name without the csv extension</li>
<li>
<code>amr_barcodes</code> : the barcodes of each same (note if you did not barcode any of your samples enter <em>none</em> in all of the cells). <strong>In the AMR workflow, barcodes are listed as “barcode” and a two digit number, no barcodes are entered as“none”</strong>
</li>
<li>
<code>wimp_filename</code> : the original amr file name without the csv extension</li>
<li>
<code>wimp_barcodes</code> : the barcodes of each same (note if you did not barcode any of your samples enter <em>NA</em> in all of the cells). <strong>In the AMR workflow, barcodes are listed as “BC” and a two digit number, no barcodes are entered as“NA”</strong>
</li>
<li>
<code>additionally information</code> after these four required columns, you may include any additional metadata that is import, such as treatment type, sample numbers, etc.</li>
</ul>
<p>An example of a combo metadata file:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">epi2me.metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"../data/example_metadata.csv"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">epi2me.metadata</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##   arma_filename wimp_filename amra_barcode wimp_barcode Samplename  treatment</span>
<span class="co">## 1   arma_288715   324212_1777    barcode02    barcode02    Animal1    control</span>
<span class="co">## 2   arma_288715   324212_1777    barcode03    barcode03    Animal2 antibiotic</span>
<span class="co">## 3   arma_288715   324212_1777    barcode04    barcode04    Animal3    control</span>
<span class="co">## 4   arma_288716   324212_1778         none         &lt;NA&gt;    Animal4 antibiotic</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="full-automated-data-import">Full Automated data import<a class="anchor" aria-label="anchor" href="#full-automated-data-import"></a>
</h2>
<p>For both AMR and WIMP data, the raw csv’s downloaded from the epi2mw website need to be in their own directory (without any other files). Note that if you are processing both WIMP and ARM data you will need <strong>two</strong> directories one for each set of data.</p>
<div class="section level3">
<h3 id="amr-data">AMR data<a class="anchor" aria-label="anchor" href="#amr-data"></a>
</h3>
<div class="section level4">
<h4 id="amr_raw_to_phyloseq">amr_raw_to_phyloseq<a class="anchor" aria-label="anchor" href="#amr_raw_to_phyloseq"></a>
</h4>
<p>The read in for AMR data requires a directory and a metadata file. The directory will be one that only has the csv files generated from EPI2ME in it. And example of the metadata file is above. The data we will be using is from an example run on the EPI2ME pipeline. There are four options:</p>
<ul>
<li>
<code>path.to.amr.files</code> <em>required</em>: the path to the raw csv files (i.e. “Desktop/raw_data/”)</li>
<li>
<code>metadata</code> <em>required</em>: metadata following format above as a data.frame</li>
<li>
<code>coveragenumber</code> <em>optional</em> : the total length of the gene that must be present for it to be included in the count table, the default is 80%; this argument takes any number from 1 to 99. Default is <code>80</code>.</li>
<li>
<code>keepSNP</code> <em>optional</em> : if gene that are considered resistance only with a SNP mutation. Default is <code>FALSE</code> (does not include these genes) but can be changed to <code>TRUE</code> to include these genes.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps.amr.object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amr_raw_to_phyloseq.html">amr_raw_to_phyloseq</a></span><span class="op">(</span>path.to.amr.files <span class="op">=</span> <span class="st">"../data/example_amr_data/"</span>,
                                     metadata <span class="op">=</span> <span class="va">epi2me.metadata</span>,
                                     coveragenumber <span class="op">=</span> <span class="fl">80</span>, 
                                     keepSNP <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## phyloseq-class experiment-level object</span>
<span class="co">## otu_table()   OTU Table:         [ 67 taxa and 4 samples ]</span>
<span class="co">## sample_data() Sample Data:       [ 4 samples by 7 sample variables ]</span>
<span class="co">## tax_table()   Taxonomy Table:    [ 67 taxa by 4 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level4">
<h4 id="amr_raw_to_metagenomeseq">amr_raw_to_metagenomeseq<a class="anchor" aria-label="anchor" href="#amr_raw_to_metagenomeseq"></a>
</h4>
<p>The same parameters for importing to metagenomseq are used as those that were used above in the <code>amr_raw_to_phyloseq</code> function:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mgs.amr.object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amr_raw_to_metagenomeseq.html">amr_raw_to_metagenomeseq</a></span><span class="op">(</span>path.to.amr.files <span class="op">=</span> <span class="st">"../data/example_amr_data/"</span>,
                                     metadata <span class="op">=</span> <span class="va">epi2me.metadata</span>,
                                     coveragenumber <span class="op">=</span> <span class="fl">80</span>, 
                                     keepSNP <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required namespace: metagenomeSeq</span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mgs.amr.object</span></code></pre></div>
<pre><code><span class="co">## MRexperiment (storageMode: environment)</span>
<span class="co">## assayData: 67 features, 4 samples </span>
<span class="co">##   element names: counts </span>
<span class="co">## protocolData: none</span>
<span class="co">## phenoData</span>
<span class="co">##   sampleNames: arma_288715_barcode02 arma_288715_barcode03</span>
<span class="co">##     arma_288715_barcode04 arma_288716_none</span>
<span class="co">##   varLabels: arma_filename wimp_filename ... sampleID (7 total)</span>
<span class="co">##   varMetadata: labelDescription</span>
<span class="co">## featureData</span>
<span class="co">##   featureNames: 36033 36036 ... 40739 (67 total)</span>
<span class="co">##   fvarLabels: OTUname Drug.Class ... ARO.Name (5 total)</span>
<span class="co">##   fvarMetadata: labelDescription</span>
<span class="co">## experimentData: use 'experimentData(object)'</span>
<span class="co">## Annotation:</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="wimp">WIMP<a class="anchor" aria-label="anchor" href="#wimp"></a>
</h3>
<div class="section level4">
<h4 id="wimp_raw_to_phyloseq">wimp_raw_to_phyloseq<a class="anchor" aria-label="anchor" href="#wimp_raw_to_phyloseq"></a>
</h4>
<p>WIMP files used the same idea as the AMR files but use the package <a href="https://github.com/sherrillmix/taxonomizr" class="external-link">taxonomizr</a> to help add taxonomic hierarchical information.</p>
<p>The read in for WIMP data requires a directory and a metadata file. The directory will be one that only has the csv files generated from EPI2ME in it. And example of the metadata file is above. The data we will be using is from an example run on the EPI2ME pipeline. There are four options:</p>
<ul>
<li>
<code>path.to.wimp.files</code> <em>required</em>: the path to the raw csv files (i.e. “Desktop/raw_data/”)</li>
<li>
<code>metadata</code> <em>required</em>: metadata following format above as a data.frame</li>
<li>
<code>keep.unclassifed</code> <em>optional</em> : keep genes that do not classify or do not classify beyond a superkingdom. Default is <code>FALSE</code> (does not include these reads) but can be changed to <code>TRUE</code> to include these reads</li>
<li>
<code>keep.human</code> <em>optional</em> : Keep reads associated with homo sapien (for microbiome usually considered a contaminant) Default is <code>FALSE</code> (does not include human associated reads) but can be changed to <code>TRUE</code> to include these reads.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps.wimp.object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wimp_raw_to_phyloseq.html">wimp_raw_to_phyloseq</a></span><span class="op">(</span>path.to.wimp.files <span class="op">=</span> <span class="st">"../data/example_wimp_data/"</span>,
                                     metadata <span class="op">=</span> <span class="va">epi2me.metadata</span>,
                                     keep.unclassifed <span class="op">=</span> <span class="cn">FALSE</span>, 
                                     keep.human <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="wimp_raw_to_metagenomeseq">wimp_raw_to_metagenomeseq<a class="anchor" aria-label="anchor" href="#wimp_raw_to_metagenomeseq"></a>
</h4>
<p>Like the functions for AMR, the same parameters for importing to metagenomseq are used as those that were used above in the <code>wimp_raw_to_phyloseq</code> function:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mgs.wimp.object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wimp_raw_to_metagenomeseq.html">wimp_raw_to_metagenomeseq</a></span><span class="op">(</span>path.to.wimp.files <span class="op">=</span> <span class="st">"../data/example_wimp_data/"</span>,
                                     metadata <span class="op">=</span> <span class="va">epi2me.metadata</span>,
                                     keep.unclassifed <span class="op">=</span> <span class="cn">FALSE</span>, 
                                     keep.human <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">mgs.wimp.object</span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="step-by-step-import">Step-by-step import<a class="anchor" aria-label="anchor" href="#step-by-step-import"></a>
</h2>
<p>In some cases you might not want a phyloseq or metagenomeseq object but instead just a count matrix or taxonomic list. In these cases you can use the below functions.</p>
<div class="section level3">
<h3 id="amr-data-1">AMR data<a class="anchor" aria-label="anchor" href="#amr-data-1"></a>
</h3>
<div class="section level4">
<h4 id="read_in_amr_file">read_in_amr_file<a class="anchor" aria-label="anchor" href="#read_in_amr_file"></a>
</h4>
<p>This takes the directory the AMR csv files are in and creates a count matrix that can be used in down stream analysis. The inputs are similar to those in the previous examples (but does not require metadata):</p>
<ul>
<li>
<code>path.to.amr.files</code> <em>required</em>: the path to the raw csv files (i.e. “Desktop/raw_data/”)</li>
<li>
<code>coveragenumber</code> <em>optional</em> : the total length of the gene that must be present for it to be included in the count table, the default is 80%; this argument takes any number from 1 to 99. Default is <code>80</code>.</li>
<li>
<code>keepSNP</code> <em>optional</em> : if gene that are considered resistance only with a SNP mutation. Default is <code>FALSE</code> (does not include these genes) but can be changed to <code>TRUE</code> to include these genes.</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">amr.count.table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_in_amr_files.html">read_in_amr_files</a></span><span class="op">(</span>path.to.amr.files <span class="op">=</span> <span class="st">"../data/example_amr_data/"</span>,
                                     coveragenumber <span class="op">=</span> <span class="fl">80</span>, 
                                     keepSNP <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">amr.count.table</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    CVTERMID arma_288715_barcode02 arma_288715_barcode03 arma_288715_barcode04</span>
<span class="co">## 1:    36033                     5                     1                     1</span>
<span class="co">## 2:    36036                     0                     0                     1</span>
<span class="co">## 3:    36213                     0                     1                     1</span>
<span class="co">## 4:    36355                     3                     2                     3</span>
<span class="co">## 5:    36376                     0                     0                     0</span>
<span class="co">## 6:    36448                     2                     0                     0</span>
<span class="co">##    arma_288716_none</span>
<span class="co">## 1:                3</span>
<span class="co">## 2:                0</span>
<span class="co">## 3:                1</span>
<span class="co">## 4:                1</span>
<span class="co">## 5:                1</span>
<span class="co">## 6:                1</span></code></pre>
</div>
<div class="section level4">
<h4 id="generate_amr_taxonomy">generate_amr_taxonomy<a class="anchor" aria-label="anchor" href="#generate_amr_taxonomy"></a>
</h4>
<p>This function assigns AMR taxonomic hierarchical information from <a href="https://card.mcmaster.ca/" class="external-link">CARD</a> using a count table with CV TERM ID’s as the first column (CVTERMID). Only one input is needed:</p>
<ul>
<li>
<code>amr.count.table</code> <em>required</em>: data frame of generated with <code>amr.count.table</code> or that has <code>CVTERMID</code> as the first column for AMR taxonomic assignment</li>
<li>
<code>verbose</code> <em>optional</em> : only a subset of coumn names are included in the output by default: (<code>CVTERMID</code>,<code>Drug Class</code>, <code>AMR Gene Family</code>, <code>Resistance Mechanism</code>, and <code>ARO Name</code>) if <code>verbose=TRUE</code> 13 columns are returned.</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">amr.taxonomy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_amr_taxonomy.html">generate_amr_taxonomy</a></span><span class="op">(</span>amr.count.table <span class="op">=</span> <span class="va">amr.count.table</span>,
                                         verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">amr.taxonomy</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    CVTERMID</span>
<span class="co">## 1:    36033</span>
<span class="co">## 2:    36036</span>
<span class="co">## 3:    36213</span>
<span class="co">## 4:    36355</span>
<span class="co">## 5:    36376</span>
<span class="co">## 6:    36448</span>
<span class="co">##                                                                                                                                                                                                                                                        Drug Class</span>
<span class="co">## 1:                                                                                                                                                                                                                                     fluoroquinolone antibiotic</span>
<span class="co">## 2:                                                                                                                                                                                                                                     fluoroquinolone antibiotic</span>
<span class="co">## 3:                                                                                                                                                                                                                                     fluoroquinolone antibiotic</span>
<span class="co">## 4:                                                                                                                        cephalosporin;fluoroquinolone antibiotic;glycylcycline;penam;phenicol antibiotic;rifamycin antibiotic;tetracycline antibiotic;triclosan</span>
<span class="co">## 5: aminocoumarin antibiotic;aminoglycoside antibiotic;carbapenem;cephalosporin;cephamycin;fluoroquinolone antibiotic;glycylcycline;macrolide antibiotic;penam;penem;peptide antibiotic;phenicol antibiotic;rifamycin antibiotic;tetracycline antibiotic;triclosan</span>
<span class="co">## 6:                                                                                                                                                                                                                                                            N/A</span>
<span class="co">##                                                                                                                                                                  AMR Gene Family</span>
<span class="co">## 1:                                                                                                                             ATP-binding cassette (ABC) antibiotic efflux pump</span>
<span class="co">## 2:                                                                                                                    major facilitator superfamily (MFS) antibiotic efflux pump</span>
<span class="co">## 3:                                                                                                                    major facilitator superfamily (MFS) antibiotic efflux pump</span>
<span class="co">## 4:                                                                                                              resistance-nodulation-cell division (RND) antibiotic efflux pump</span>
<span class="co">## 5: ATP-binding cassette (ABC) antibiotic efflux pump;major facilitator superfamily (MFS) antibiotic efflux pump;resistance-nodulation-cell division (RND) antibiotic efflux pump</span>
<span class="co">## 6:                                                                                                                    major facilitator superfamily (MFS) antibiotic efflux pump</span>
<span class="co">##    Resistance Mechanism ARO Name</span>
<span class="co">## 1:    antibiotic efflux     patA</span>
<span class="co">## 2:    antibiotic efflux     emrA</span>
<span class="co">## 3:    antibiotic efflux     emrB</span>
<span class="co">## 4:    antibiotic efflux     acrB</span>
<span class="co">## 5:    antibiotic efflux     TolC</span>
<span class="co">## 6:    antibiotic efflux     emrD</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="wimp-data">WIMP data<a class="anchor" aria-label="anchor" href="#wimp-data"></a>
</h3>
<div class="section level4">
<h4 id="read_in_wimp_file">read_in_wimp_file<a class="anchor" aria-label="anchor" href="#read_in_wimp_file"></a>
</h4>
<p>This takes the directory the WIMP csv files are in and creates a count matrix that can be used in down stream analysis. The inputs are similar to those in the previous examples (but does not require metadata):</p>
<ul>
<li>
<code>path.to.wimp.files</code> <em>required</em>: the path to the raw csv files (i.e. “Desktop/raw_data/”)</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wimp.count.table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_in_wimp_files.html">read_in_wimp_files</a></span><span class="op">(</span>path.to.wimp.files <span class="op">=</span> <span class="st">"../data/example_wimp_data/"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Reading in raw files from ../data/example_wimp_data/</span></code></pre>
<pre><code><span class="co">## The percentage of classifed reads was 94.4</span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">wimp.count.table</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    taxID 324212_1777_barcode02 324212_1777_barcode03 324212_1777_barcode04</span>
<span class="co">## 1:     0                     9                     6                     6</span>
<span class="co">## 2:   286                     3                     0                     0</span>
<span class="co">## 3:   287                     8                     9                    11</span>
<span class="co">## 4:   543                     2                     1                     2</span>
<span class="co">## 5:   544                     0                     0                     1</span>
<span class="co">## 6:   561                     0                     0                     0</span>
<span class="co">##    324212_1778_NA</span>
<span class="co">## 1:              7</span>
<span class="co">## 2:              1</span>
<span class="co">## 3:             16</span>
<span class="co">## 4:              0</span>
<span class="co">## 5:              0</span>
<span class="co">## 6:              1</span></code></pre>
</div>
<div class="section level4">
<h4 id="generate_wimp_taxonomy">generate_wimp_taxonomy<a class="anchor" aria-label="anchor" href="#generate_wimp_taxonomy"></a>
</h4>
<p>This function assigns phylogenetic taxonomic hierarchical information with the help of <code>taxonomizr</code> a count table with NCBI taxonomic ID’s (<code>taxID</code> as the first column is required.</p>
<ul>
<li>
<code>wimp.count.table</code> <em>required</em>: data frame of generated with <code>wimp.count.table</code> or that has <code>taxID</code> as the first column for phylogenetic taxonomic assignment</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wimp.taxonomy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_wimp_taxonomy.html">generate_wimp_taxonomy</a></span><span class="op">(</span>wimp.count.table <span class="op">=</span> <span class="va">wimp.count.table</span><span class="op">)</span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Maggie Weinroth.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
